src/Juvix/Core/IR/Types.hs

```diff
@@ -1,58 +1,139 @@
+{-# LANGUAGE UndecidableInstances #-}
+
 -- | Quantitative type implementation inspired by
 --   Atkey 2018 and McBride 2016.
-module Juvix.Core.IR.Types where
+module Juvix.Core.IR.Types
+  (module Juvix.Core.IR.Types,
+   module Juvix.Core.IR.Extension)
+where
 
 import Juvix.Core.Usage
+import Juvix.Core.IR.Extension
 import Juvix.Library hiding (show)
 import Prelude (Show (..), String)
 
+
 -- | checkable terms
-data Term primTy primVal
+data Term' ext primTy primVal
   = -- | (sort i) i th ordering of (closed) universe.
-    Star Natural
+    Star' Natural (XStar ext primTy primVal)
   | -- | 'PrimTy' primitive type
-    PrimTy primTy
+    PrimTy' primTy (XPrimTy ext primTy primVal)
   | -- | formation rule of the dependent function type 'PI'.
     -- the Usage(π) tracks how many times x is used.
-    Pi Usage (Term primTy primVal) (Term primTy primVal)
+    Pi' Usage (Term' ext primTy primVal) (Term' ext primTy primVal)
+       (XPi ext primTy primVal)
   | -- | 'LAM' Introduction rule of PI.
     -- The abstracted variable's usage is tracked with the Usage(π).
-    Lam (Term primTy primVal)
+    Lam' (Term' ext primTy primVal) (XLam ext primTy primVal)
   | -- | 'CONV' conversion rule. TODO make sure 0Γ ⊢ S≡T
     -- 'Elim' is the constructor that embeds Elim to Term
-    Elim (Elim primTy primVal)
-  deriving (Eq)
+    Elim' (Elim' ext primTy primVal) (XElim ext primTy primVal)
+  | -- | Extension point for other construction forms
+    TermX (TermX ext primTy primVal)
+
+deriving instance
+  (Eq primTy, Eq primVal, TEAll Eq ext primTy primVal) =>
+  Eq (Term' ext primTy primVal)
+
+deriving instance
+  (Show primTy, Show primVal, TEAll Show ext primTy primVal) =>
+  Show (Term' ext primTy primVal)
 
+-- FIXME add pretty-printer
```

Comment:
Yes. We should track this separately. https://github.com/cryptiumlabs/juvix/issues/308.